packages/gatsby-source-drupal/src/gatsby-node.js

```diff
@@ -114,170 +118,256 @@ exports.sourceNodes = async (
 
   drupalFetchActivity.end()
 
-  // Make list of all IDs so we can check against that when creating
-  // relationships.
-  const ids = {}
+  const nodes = new Map()
+
+  // first pass - create basic nodes
   _.each(allData, contentType => {
     if (!contentType) return
     _.each(contentType.data, datum => {
-      ids[datum.id] = true
+      const node = nodeFromData(datum, createNodeId)
+      nodes.set(node.id, node)
     })
   })
 
-  // Create back references
-  const backRefs = {}
-
-  /**
-   * Adds back reference to linked entity, so we can later
-   * add node link.
-   */
-  const addBackRef = (linkedId, sourceDatum) => {
-    if (ids[linkedId]) {
-      if (!backRefs[linkedId]) {
-        backRefs[linkedId] = []
-      }
-      backRefs[linkedId].push({
-        id: sourceDatum.id,
-        type: sourceDatum.type,
-      })
+  // second pass - handle relationships and back references
+  nodes.forEach(node => {
+    handleReferences(node, {
+      getNode: nodes.get.bind(nodes),
+      createNodeId,
+    })
+  })
+
+  reporter.info(`Downloading remote files from Drupal`)
+  downloadingFilesActivity.start()
+
+  // Download all files (await for each pool to complete to fix concurrency issues)
+  await asyncPool(
+    concurrentFileRequests,
+    [...nodes.values()].filter(isFileNode),
+    async node => {
+      await downloadFile(
+        { node, store, cache, createNode, createNodeId },
+        pluginOptions
+      )
     }
+  )
+
+  downloadingFilesActivity.end()
+
+  // Create each node
+  for (const node of nodes.values()) {
+    node.internal.contentDigest = createContentDigest(node)
+    createNode(node)
   }
+}
 
-  _.each(allData, contentType => {
-    if (!contentType) return
-    _.each(contentType.data, datum => {
-      if (datum.relationships) {
-        _.each(datum.relationships, (v, k) => {
-          if (!v.data) return
+exports.onCreateDevServer = (
+  {
+    app,
+    createNodeId,
+    getNode,
+    actions,
+    store,
+    cache,
+    createContentDigest,
+    reporter,
+  },
+  pluginOptions
+) => {
+  app.use(
+    `/___updatePreview/`,
```

Comment:
I would propose to use `/__update/gatsby-source-drupal` or something like this here, so the update endpoint is namespaced. (This would need changes in Gatsby drupal module)

Reply:
If you provide me with the namespaces I can update the Drupal module to reflect this once it's merged. I'm assuming you would want to use `___update`, `___create` and `___delete`?

Reply:
I don't have opinion on separate endpoints for update/create/delete operations - it can stay as single endpoint and operation could be send with payload (for our use case update and create will have same handler anyway - only delete will have different handling). Only thing that I'm interested in is using namespace for `gatsby-source-drupal` (as we will likely start adding similar endpoints for other plugins)

Reply:
For the sake of time. I think it's fine to merge this - this PR was in limbo for too long. I'll just add explicit note to README that there likely will be breaking changes coming that will require updating both gatsby-source-drupal and gatsby drupal module in future